### ISSUE TITLE
QC Speed Limit Layer

### ISSUE THREAD

**Issue by jovenc** [CityofToronto/bdit_data-sources#271]
The very first version of speed limit layer (2019) in `jchew.speed_limit_aug2019` contains issues as shown in #188 whereas the documentation for the QC process is [here](https://github.com/CityofToronto/bdit_data-sources/tree/text_to_centreline/gis/posted_speed_limit_update/automated/QC). Need to revisit after the 2020 speed limit layer is completed.

For v1 of 2020 speed limit layer in `jchew.bylaws_routing1`, there are a couples of bylaw_id that has to be checked. They are 

- [x] 6667 with low confidence,

- [x] 6848 where kings highway(?) is geocoded to Yonge St

- [x] 6645 where there are two parts of McGillivray Ave

**Comment by jovenc** [comment]
For bylaw_id = 6667 where `highway` = "Welland Ave"	and `between` = "Lawrence Ave W and Douglas Avenue", it's routed as shown below and that is not right.
![image](https://user-images.githubusercontent.com/54872846/75817537-45a8ec80-5d65-11ea-8002-224badaa4278.png)

I found that the reason is that one of the intersection points found is not right. Tracing back to the function `gis._get_intersection_id`, I found the levenshtein distance below where int_id 13454250 is the one being chosen but int_id 134566028 is the right one. 
![image](https://user-images.githubusercontent.com/54872846/75817716-ac2e0a80-5d65-11ea-847d-9fcff01d9a6e.png)

Running a check on the `gis.centreline` table, I found this. 
![image](https://user-images.githubusercontent.com/54872846/75817896-f57e5a00-5d65-11ea-8838-cadd53e0a45b.png)

Since changing 'Lawrence Ave W' to 'Lawrence Ave E' only takes 1 step whereas changing 'Welland Ave' to 'Welland Rd' takes 3 steps, the intersection point with the less levenshtein distance is chosen instead, resulting in the segments of road that does not reflect the bylaws text.

I also found that when there are three street names in the `intersec5` column and two of the street names are almost the same with the slight change of E to W or vice versa (which happen particularly frequently for Yonge St like `Adelaide St E / Yonge St / Adelaide St W`, `King St E / Yonge St / King St W`, `Bloor St E / Yonge St / Bloor St W` etc), the intersection points found is weird because of the reason above.

**Comment by jovenc** [comment]
For bylaw_id = 6848 where highway = "Dundas Street West" and between = "King's Highway 27 and East Mall Crescent" , it's routed ALL THE WAY TO YONGE ST AGAIN as shown below and that is not right. (highlighted is the road segment we want)
![image](https://user-images.githubusercontent.com/54872846/75819292-67f03980-5d68-11ea-9bf9-010336fe2326.png)

Since there's nth similar to 'King's Highway 27' with levenshtein distance less than 4, this happened as changing 'Dundas St W' to 'Dundas St E' only takes 1 step.
![image](https://user-images.githubusercontent.com/54872846/75819419-a554c700-5d68-11ea-942c-6570789351e6.png)

**Comment by jovenc** [comment]
For bylaw_id = 6645 where highway = "McGillivray Ave" and between = "The west end of McGillivray Ave and Kelso Ave", it's routed as shown below and that is not right. (highlighted is the road segment we want) 
![image](https://user-images.githubusercontent.com/54872846/75820184-e4374c80-5d69-11ea-8360-2e1725106ef9.png)

This happen because there are two 'McGillivray Ave' in Toronto (highlighted in yellow and maroon above) and they both are cul de sac as shown below. int_id 13452712 is the right one but 13453328 is the one return from the function `gis._get_intersection_id`
![image](https://user-images.githubusercontent.com/54872846/75820771-f4036080-5d6a-11ea-9b7e-29ad2acb852a.png)

**Comment by jovenc** [comment]
`gis._get_intersection_id` looks like this 
```
CREATE OR REPLACE FUNCTION gis._get_intersection_id(highway2 TEXT, btwn TEXT, not_int_id INT)
RETURNS INT[] AS $$
DECLARE
oid INT;
lev_sum INT;
int_id_found INT;

BEGIN
SELECT intersections.objectid, 

SUM(LEAST
(levenshtein(TRIM(intersections.street), TRIM(highway2), 1, 1, 1)
, levenshtein(TRIM(intersections.street), TRIM(btwn), 1, 1, 1)))

, intersections.int_id
INTO oid, lev_sum, int_id_found
FROM
(gis.centreline_intersection_streets --to get street name
LEFT JOIN gis.centreline_intersection --to get int_id 
USING(objectid)) AS intersections

WHERE (levenshtein(TRIM(intersections.street), TRIM(highway2), 1, 1, 1) < 4 
OR levenshtein(TRIM(intersections.street), TRIM(btwn), 1, 1, 1) < 4) 
AND intersections.int_id  <> not_int_id

GROUP BY intersections.objectid, intersections.int_id
HAVING COUNT(DISTINCT TRIM(intersections.street)) > 1
ORDER BY AVG(LEAST(levenshtein(TRIM(intersections.street), TRIM(highway2), 1, 1, 1), levenshtein(TRIM(intersections.street),  TRIM(btwn), 1, 1, 1)))

LIMIT 1;

raise notice 'highway2 being matched: % btwn being matched: % not_int_id: % intersection arr: %', highway2, btwn, not_int_id, ARRAY[oid, lev_sum];

RETURN ARRAY[oid, lev_sum, int_id_found];

END;
$$ LANGUAGE plpgsql;
```
But I only use the SELECT block for testing purposes and removed `LIMIT 1` to see all results returned which explains the screengrabs above with `levenshtein_dist` column
```
SELECT intersections.objectid, 

SUM(LEAST
(levenshtein(TRIM(intersections.street), TRIM('Welland Ave'), 1, 1, 1)
, levenshtein(TRIM(intersections.street), TRIM('Lawrence Ave W'), 1, 1, 1))) AS levenshtein_dist

, intersections.int_id
--INTO oid, lev_sum, int_id_found
FROM
(gis.centreline_intersection_streets --to get street name
LEFT JOIN gis.centreline_intersection --to get int_id 
USING(objectid)) AS intersections

WHERE (levenshtein(TRIM(intersections.street), TRIM('Welland Ave'), 1, 1, 1) < 4 
OR levenshtein(TRIM(intersections.street), TRIM('Lawrence Ave W'), 1, 1, 1) < 4) 
--AND intersections.int_id  <> not_int_id

GROUP BY intersections.objectid, intersections.int_id
HAVING COUNT(DISTINCT TRIM(intersections.street)) > 1
ORDER BY AVG(LEAST(levenshtein(TRIM(intersections.street), TRIM('Welland Ave'), 1, 1, 1), 
				   levenshtein(TRIM(intersections.street),  TRIM('Lawrence Ave W'), 1, 1, 1)))

--LIMIT 1;
```

**Comment by jovenc** [comment]
Based on the work from #270 , the latest version of `jchew.text_to_centreline_updated()` can now deal with

1. normal cases -- two intersections without any translated point
2. entire length
3. an intersection and an offset (when offset is at the second part aka after "and")
4. two intersections with at least one offset

However, I have found a few newer issues.
a) There's another special case where there's only one intersection (btwn1 = btwn2) but two offsets (metres_btwn1 IS NOT NULL & metres_btwn2 IS NOT NULL)
_(as well as 3 of them WHERE btwn1 = btwn2 AND metres_btwn1 IS NOT NULL BUT metres_btwn2 IS NULL)_
![image](https://user-images.githubusercontent.com/54872846/77802945-60177280-7052-11ea-85de-aae1132d5786.png)

b) For bylaw_id = 5266, highway = 'Antler Street', and between = 'Franklin Avenue and a point 100 metres west', this description is not possible if you look at the google map. The highlighted part is only 50m long and it is east of the intersection.
![image](https://user-images.githubusercontent.com/54872846/77803054-a8369500-7052-11ea-8fbb-54b775a1bc7c.png)
![image](https://user-images.githubusercontent.com/54872846/77803082-b2f12a00-7052-11ea-9535-c641171cc4bb.png)

c) For bylaw_id = 5244, highway = 'Renforth Drive', and between = 'A point 300 metres north of Eglinton Avenue West and a point 250 metres south west of Silver Dart Drive', some of the centrelines are missing as shown below. Centrelines in the highlighted part is missing.
![image](https://user-images.githubusercontent.com/54872846/77803306-327ef900-7053-11ea-9e89-4291ac43ad5f.png)

d) For bylaw_id = 6842 , highway='Black Creek Drive(southbound)' and between = 'Eglinton Avenue West and a point 200 metres north of Weston Road',
```
SELECT * FROM jchew.text_to_centreline_updated
(6842, 'Black Creek Drive (southbound)', 'Eglinton Avenue West and a point 200 metres north of Weston Road', NULL);
```
does not yield any results but the below does
```
SELECT * FROM jchew.text_to_centreline_updated
(6842, 'Black Creek Drive(southbound)', 'Eglinton Avenue West and a point 200 metres north of Weston Road', NULL);
```
The only difference is the space between "Black Creek Drive" and "(southbound)".

e) Again for bylaw_id = 6842, one of the results return a line_geom that's not viewable. 
```
SELECT * FROM jchew.text_to_centreline_updated
(6842, 'Black Creek Drive(southbound)', 'Eglinton Avenue West and a point 200 metres north of Weston Road', NULL);
```
The query above results in 3 centrelines which have geo_id of 8548, 8675 and 8755. The one with geo_id = 8755 is the problematic one. I checked `gis.centreline` and it looks find but it is something weird in my result. Might also be that centreline is really close to the centreline I want? (Not too sure why)
